### PR TITLE
feat: implement BEP-333(BNB Chain Fusion)

### PR DESCRIPTION
**Pull request by j75689**
### Description

Implement the functions for the `BC-Fusion` plan.

The Beacon Chain in the `BC-Fusion` plan consists of four stages:

1. **FirstSunsetFork - Disable Certain Message Types:**
   In this stage, the system will disable specific message types to prevent users from creating additional assets on the Beacon Chain. The following `Msgs` will be disabled:
   - TimeLockMsg
   - TimeRelockMsg
   - FreezeMsg
   - IssueMsg 
   - MintMsg 
   - IssueMiniMsg 
   - HTLTMsg 
   - DepositHTLTMsg
   - MsgSideChainDelegate
   - MsgSideChainRedelegate
   - MsgCreateSideChainValidator
   - MsgCreateSideChainValidatorWithVoteAddr
   - MsgEditSideChainValidator
   - MsgEditSideChainValidatorWithVoteAddr

   `MsgSideChainStakeMigration` will be enabled to allow users to redelegate tokens to BSC.        
   `MsgSideChainSubmitProposal` will be disabled after the SideChain Total Voting Power falls below 5M BNB.

2. **SecondSunsetFork - Token Refunding:**
   During this stage, the system will process token refunds for users who have assets locked on the Beacon Chain. Refunds will be initiated for TimeLock, AtomicSwap, and `Delegation` on the side chain. `MsgSideChainUndelegate` on the Beacon Chain will be disabled to prevent undelegation and ensure system efficiency.

3. **FinalSunsetFork - Closure of Cross-Chain Channels:**
   This marks the final stage of `BC-Fusion`. The system will close all cross-chain channels. After the `FinalSunsetFork`, the system will no longer support cross-chain transactions.

4. **Token Migration:**
   After the `FinalSunsetFork`, users may still have assets on the Beacon Chain. The system will support token migration, allowing users to move their assets to BSC. For detailed instructions, please refer to [BEP-299](https://github.com/bnb-chain/BEPs/pull/299).

### Rationale

For the reasoning behind this plan, refer to [BEP-333](https://github.com/bnb-chain/BEPs/pull/333).

### Example

n/a

### Notable Changes

* Tokens
* Staking
* Governance
* IBC

### CODE REVIEW

**Comment by hazemdraosha111** [comment]
11

File: plugins/migrate/client/cli/tx.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/tendermint/tendermint/crypto/secp256k1"
+
+	"github.com/bnb-chain/node/plugins/migrate"
+)
+
+const (
+	flagBSCOperatorAddress = "bsc-operator-address"
+)
+
+func SignValidatorOwnerShipCmd(cdc *codec.Codec) *cobra.Command {
```
Comment (forcodedancing) [inline plugins/migrate/client/cli/tx.go:29]: This just prints the sign data. How can we get the binding relationship between the old and the new validators?

File: plugins/recover/client/cli/commands.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	"github.com/spf13/cobra"
+)
+
+func AddCommands(cmd *cobra.Command, cdc *codec.Codec) {
+	airdropCmd := &cobra.Command{
+		Use:   "recover",
+		Short: "recover commands",
+	}
+
+	airdropCmd.AddCommand(
```
Comment (forcodedancing) [inline plugins/recover/client/cli/commands.go:15]: Rename `airdropCmd`

File: plugins/tokens/plugin.go [hunk 3]
```diff
@@ -51,6 +51,76 @@ func createQueryHandler(mapper Mapper, queryPrefix string) app.AbciQueryHandler
...
+	if !sdk.IsUpgrade(sdk.SecondSunsetFork) {
+		return
+	}
+	logger := bnclog.With("module", "tokens")
+	logger.Info("unlock the time locks", "blockHeight", ctx.BlockHeight())
+
+	iterator := timelockKeeper.GetTimeLockRecordIterator(ctx)
+	defer iterator.Close()
+	i := 0
+	failedCount := 0
```
Comment (forcodedancing) [inline plugins/tokens/plugin.go:68]: In cosmos-sdk, we also have a limit for failures. Do we also need such a strategy here?

File: plugins/migrate/msg.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// Type implements types.Msg.
+func (ValidatorOwnerShip) Type() string {
+	return MsgType
+}
+
+// ValidateBasic implements types.Msg.
+func (msg ValidatorOwnerShip) ValidateBasic() sdk.Error {
+	return nil
```
[the remark this answers is not in the document]
Comment (j75689) [reply plugins/migrate/msg.go:74]: It's converted by `common.HexToAddress`, so it should be valid.

image

File: plugins/tokens/plugin.go (hunk 3 above)
Comment (j75689) [reply plugins/tokens/plugin.go:68]: `TimeUnlock` is a very simple function. It shouldn't fail.
It failed and I think we need to do manual analysis

File: plugins/recover/client/cli/commands.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/spf13/cobra"
+)
+
+func AddCommands(cmd *cobra.Command, cdc *codec.Codec) {
+	airdropCmd := &cobra.Command{
+		Use:   "recover",
+		Short: "recover commands",
```
[the remark this answers is not in the document]
Comment (j75689) [reply plugins/recover/client/cli/commands.go:12]: accept

File: plugins/recover/client/cli/commands.go (hunk 2 above)
Comment (j75689) [reply plugins/recover/client/cli/commands.go:15]: accept

**Review by unclezoro — APPROVED** [review]

**Pull request merged**